### PR TITLE
make docker-credential-wincred work like docker-credential-osxkeychain

### PR DESCRIPTION
**Pull request by ekcasey**
* fetch credentials for server with matching hostname if scheme, path, or port are not provided
* if the credential request includes specific scheme, path, or port that does not match entry, don't return
* extract url helpers into a package

Signed-off-by: Danny Joyce <djoyce@pivotal.io>
Signed-off-by: Emily Casey <ecasey@pivotal.io>

### CODE REVIEW

**Comment by ekcasey** [comment]
This fixes issues fetching docker credentials in libraries like https://github.com/google/go-containerregistry that resolve Docker Hub registry url differently (https://index.docker.io vs https://index.docker.io/v1)

**Comment by ekcasey** [comment]
@vdemeester It looks like the Travis build is broken for reasons unrelated to our changes. If that is not a correct diagnosis, let @djoyahoy and I know so we fix it :)

**Comment by guillaumerose** [comment]
CI is now fixed. Can you rebase ? Thanks

**Comment by ekcasey** [comment]
@guillaumerose done!

File: osxkeychain/osxkeychain_darwin.go [hunk 1]
```diff
@@ -10,9 +10,8 @@ package osxkeychain
 import "C"
 import (
 	"errors"
-	"net/url"
+	"github.com/docker/docker-credential-helpers/registryurl"
```
Comment (silvin-lubecki) [inline osxkeychain/osxkeychain_darwin.go]: nit picking: Can you re-order the imports? `goimports` should do that for you 😺

File: wincred/wincred_windows.go [hunk 2]
```diff
@@ -2,6 +2,8 @@ package wincred
 
 import (
 	"bytes"
+	"github.com/docker/docker-credential-helpers/registryurl"
```
Comment (silvin-lubecki) [inline wincred/wincred_windows.go]: nit: please re-organize the imports here too

File: wincred/wincred_windows.go [hunk 3]
```diff
@@ -51,6 +61,72 @@ func (h Wincred) Get(serverURL string) (string, string, error) {
 	return "", "", credentials.NewErrCredentialsNotFound()
 }
 
+func getTarget(serverURL string) (string, error) {
+	s, err := registryurl.Parse(serverURL)
+	if err != nil {
+		return serverURL, nil
```
Comment (silvin-lubecki) [inline wincred/wincred_windows.go]: nit: return empty string here instead?

File: wincred/wincred_windows.go [hunk 4]
```diff
@@ -51,6 +61,72 @@ func (h Wincred) Get(serverURL string) (string, string, error) {
...
+	}
+
+	creds, err := winc.List()
+	if err != nil {
+		return "", err
+	}
+
+	targets := make([]string, 0)
```
Comment (silvin-lubecki) [inline wincred/wincred_windows.go]: nit: `var targets []string`

File: wincred/wincred_windows.go [hunk 5]
```diff
@@ -51,6 +61,72 @@ func (h Wincred) Get(serverURL string) (string, string, error) {
...
+	for i := range creds {
+		attrs := creds[i].Attributes
+		for _, attr := range attrs {
+			if strings.Compare(attr.Keyword, "label") == 0 &&
```
Comment (silvin-lubecki) [inline wincred/wincred_windows.go]: From the documentation: https://golang.org/pkg/strings/#Compare
> Compare is included only for symmetry with package bytes. It is usually clearer and always faster to use the built-in string comparison operators ==, <, >, and so on.

So I guess `if attr.Keyword == "label"` is simpler 😸

File: wincred/wincred_windows.go [hunk 6]
```diff
@@ -51,6 +61,72 @@ func (h Wincred) Get(serverURL string) (string, string, error) {
...
+				bytes.Compare(attr.Value, []byte(credentials.CredsLabel)) == 0 {
```
Comment (silvin-lubecki) [inline wincred/wincred_windows.go]: You should use [bytes.Equal](https://golang.org/pkg/bytes/#Equal) instead.
What is the real type behind `attr.Value`? Is it an arbitrary byte slice? Or always a string encoded into bytes? My point is: even if the result is the same, I think it's more clearer to cast `attr.Value` as string instead.

File: wincred/wincred_windows.go (hunk 3 above)
Comment (ekcasey) [reply wincred/wincred_windows.go]: We returned the serverURL here b/c before we made our changes there was never a requirement that the url had to parse successfully (you could still fetch an exact match from the cred store). Happy to change this if you want to make this a requirement.

Comment (silvin-lubecki) [reply wincred/wincred_windows.go]: Oh right read it too fast, you return a nil error. That's fine then 👍

File: wincred/wincred_windows.go (hunk 6 above)
Comment (ekcasey) [reply wincred/wincred_windows.go]: This piece of code existed before our PR so weren't sure if `attr.Value` needed to be a byte slice for a specific reason, happy to make the changes though.

File: wincred/wincred_windows.go [hunk 7]
```diff
@@ -51,6 +61,72 @@ func (h Wincred) Get(serverURL string) (string, string, error) {
...
+				targets = append(targets, creds[i].TargetName)
+			}
+		}
+	}
+
+	if target, found := findMatch(s, targets, exactMatch); found {
```
Comment (silvin-lubecki) [inline wincred/wincred_windows.go]: Is this really needed? Can't we just match with approximation? What is the targets length ?

Comment (ekcasey) [reply wincred/wincred_windows.go]: We assumed that if there were both an exact and an approximate match in the cred store, the exact match would be preferable. We can take the exact match out if you disagree.

Comment (silvin-lubecki) [reply wincred/wincred_windows.go]: Sounds good 😄 Ok then 👍

**Comment by ekcasey** [comment]
Thanks everybody!

**Pull request merged**